src/shared/hooks/use-linked-interactives-authoring.tsx

```diff
@@ -0,0 +1,106 @@
+import { useEffect, useState } from "react";
+import deepmerge from "deepmerge";
+import { JSONSchema6 } from "json-schema";
+import {
+  useAuthoredState, useInitMessage, getInteractiveList, ILinkedInteractive, setLinkedInteractives
+} from "@concord-consortium/lara-interactive-api";
+
+export interface ILinkedInteractiveProp {
+  label: string;
+  supportsSnapshots?: boolean;
+}
+
+export interface IProps {
+  linkedInteractiveProps?: ILinkedInteractiveProp[];
+  schema: JSONSchema6;
+}
+
+type AuthoredState = {[key: string]: any};
+
+const emptyArray: ILinkedInteractive[] = [];
+
+export const useLinkedInteractivesAuthoring = ({ linkedInteractiveProps, schema }: IProps) => {
+  const initMessage = useInitMessage<AuthoredState>();
+  const [ interactiveList, setInteractiveList ] = useState<{[label: string]: {names: string[], ids: string[]}}>({});
+  const [ cachedLinkedInteractives, setCachedLinkedInteractives ] = useState<ILinkedInteractive[]>();
+  const { authoredState } = useAuthoredState<AuthoredState>();
+
+  // Note that initMessage.linkedInteractives is never updated, even after setLinkedInteractives is sent.
+  // So, cachedLinkedInteractives is used to keep the most recent value after updates.
+  const currentLinkedInteractives = cachedLinkedInteractives || (initMessage?.mode === "authoring" && initMessage?.linkedInteractives) || emptyArray;
+
+  // Handle authored state update. Each time one of the linked interactive properties is updated, send updated
+  // linkedInteractives array to LARA parent window. cachedLinkedInteractives acts only as a cache, so we don't
+  // send the message each time the authoredState is updated (e.g. its other, unrelated fields).
+  useEffect(() => {
+    if (linkedInteractiveProps && authoredState && initMessage?.mode === "authoring") {
+      linkedInteractiveProps.forEach(li => {
+        const name = li.label;
+        const authoredStateVal = authoredState[name];
+        const linkedInteractive = currentLinkedInteractives.find(l => l.label === name);
+        if (!linkedInteractive && authoredStateVal !== undefined) {
+          // Add a new item.
+          const newArray = currentLinkedInteractives.concat({
+            id: authoredStateVal,
+            label: name
+          });
+          setCachedLinkedInteractives(newArray); // Set cached value
+          setLinkedInteractives({linkedInteractives: newArray}); // Send to LARA
+        } else if (linkedInteractive && authoredStateVal === undefined) {
+          // Remove item from the array.
+          const idx = currentLinkedInteractives.indexOf(linkedInteractive);
+          const newArray = currentLinkedInteractives.slice();
+          newArray.splice(idx, 1);
+          setCachedLinkedInteractives(newArray); // Set cached value
+          setLinkedInteractives({linkedInteractives: newArray}); // Send to LARA
+        } else if (linkedInteractive && linkedInteractive.id !== authoredStateVal) {
```

Comment:
```suggestion
        } else if (linkedInteractive?.id !== authoredStateVal) {
```

Reply:
This actually won't work here, as the logic of these two expressions is different.

`else if (linkedInteractive && linkedInteractive.id !== authoredStateVal)` will never enter the block if `linkedInteractive` is a falsy value.

`else if (linkedInteractive?.id !== authoredStateVal) {` might enter the block when `authoredStateVal` is equal to false or undefined.